src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+
+    prev->m_event = chunk->m_event;
+    prev->m_resource = chunk->m_resource;
+
+    if (prev->next)
+      prev->next->prev = prev;
+
+    UMPIRE_LOG(Debug, "Merging with prev" << prev << " and " << chunk);
+    UMPIRE_LOG(Debug, "New size: " << prev->size);
+
+    chunk->~Chunk(); // manually call destructor
+    m_chunk_pool.deallocate(chunk);
+    chunk = prev;
+  }
+
+  if (chunk->next && chunk->next->free == true) {
+    auto next = chunk->next;
+
+    UMPIRE_LOG(Debug, "Removing chunk" << next << " from free map");
+    m_free_map.erase(next->size_map_it);
+
+    chunk->size += next->size;
+    chunk->next = next->next;
+
+    chunk->m_event = next->m_event;
+    chunk->m_resource = next->m_resource;
+
+    if (chunk->next)
+      chunk->next->prev = chunk;
+
+    UMPIRE_LOG(Debug, "Merging with next" << chunk << " and " << next);
+    UMPIRE_LOG(Debug, "New size: " << chunk->size);
+
+    next->~Chunk(); // manually call destructor
+    m_chunk_pool.deallocate(next);
+  }
+
+  UMPIRE_LOG(Debug, "Inserting chunk " << chunk << " with size " << chunk->size);
+
+  if (chunk->size == chunk->chunk_size) {
+    m_releasable_blocks++;
+    m_releasable_bytes += chunk->chunk_size;
+  }
+
+  chunk->size_map_it = m_free_map.insert(std::make_pair(chunk->size, chunk));
+}
+
+void ResourceAwarePool::deallocate_resource(camp::resources::Resource r, void* ptr, std::size_t UMPIRE_UNUSED_ARG(size))
+{
+  UMPIRE_LOG(Debug, "(ptr=" << ptr << ")");
+  UMPIRE_LOG(Debug, "(Resource=" << camp::resources::to_string(r) << ")");
+  auto chunk = (*m_used_map.find(ptr)).second;
+
+  if (chunk == nullptr) {
+    UMPIRE_ERROR(runtime_error, fmt::format("The chunk can't be found! Called deallocate with ptr: {}", ptr));
+  }
+
+  auto my_r = getResource(ptr);
+  if (my_r != r) {
+    UMPIRE_ERROR(
+        runtime_error,
+        fmt::format("Called deallocate with different resource than what is returned by getResource. Called with {},",
+                    "but getResource returned: {}", camp::resources::to_string(r), camp::resources::to_string(my_r)));
+  }
+
+  // Chunk is now pending, add to list
+  m_pending_map.push_back(chunk);
+  if (m_is_coalescing == false) {
+    chunk->m_event = r.get_event();
+  }
+
+  m_used_map.erase(ptr);
+  m_current_bytes -= chunk->size;
+
+  // Call deallocate logic only for a non-pending chunk
+  if (chunk->m_event.check()) {
+    do_deallocate(chunk, ptr);
+  }
+
+  std::size_t suggested_size{m_should_coalesce(*this)};
+  if (0 != suggested_size) {
+    UMPIRE_LOG(Debug, "coalesce heuristic true, performing coalesce.");
+    do_coalesce(suggested_size);
+  }
+}
+
+void ResourceAwarePool::release()
+{
+  UMPIRE_LOG(Debug, "() " << m_free_map.size() << " chunks in free map, m_is_destructing set to " << m_is_destructing);
+
+#if defined(UMPIRE_ENABLE_BACKTRACE)
+  std::size_t prev_size{m_actual_bytes};
+#endif
+
+  // TODO:
+  // This will check all chunks in m_pending_map and erase the entry if event is complete
```

Comment:
What happens if the event is not complete? Is that a leak?

Reply:
That is a great question. If there is still a pending chunk when the `release()` is called, then it would have to wait until the pending deallocate completes in order to proceed - right? I will need to think about this more, but it may be a case for a `wait()`